File: epi_judge_cpp_solutions/delete_node_from_list.cc
#include "list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/timed_executor.h"

// Assumes node_to_delete is not tail.
void DeletionFromList(const shared_ptr<ListNode<int>>& node_to_delete) {
  node_to_delete->data = node_to_delete->next->data;
  node_to_delete->next = node_to_delete->next->next;
}

shared_ptr<ListNode<int>> DeletionFromListWrapper(
    TimedExecutor& executor, const shared_ptr<ListNode<int>>& head,
    int node_to_delete_idx) {
  shared_ptr<ListNode<int>> selected_node = head;
  while (node_to_delete_idx-- > 0) {
    if (!selected_node || !selected_node->next)
      throw std::runtime_error("Node index is out of range");
    selected_node = selected_node->next;
  }

  executor.Run([&] { DeletionFromList(selected_node); });
  return head;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "head",
                                       "node_to_delete_idx"};
  return GenericTestMain(args, "delete_node_from_list.cc",
                         "delete_node_from_list.tsv", &DeletionFromListWrapper,
                         DefaultComparator{}, param_names);
}

File: epi_judge_python_solutions/delete_node_from_list.py
import functools

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook


# Assumes node_to_delete is not tail.
def deletion_from_list(node_to_delete):

    node_to_delete.data = node_to_delete.next.data
    node_to_delete.next = node_to_delete.next.next


@enable_executor_hook
def deletion_from_list_wrapper(executor, head, node_to_delete_idx):
    node_to_delete = head
    if node_to_delete is None:
        raise RuntimeError('List is empty')
    for _ in range(node_to_delete_idx):
        if node_to_delete.next is None:
            raise RuntimeError("Can't delete last node")
        node_to_delete = node_to_delete.next

    executor.run(functools.partial(deletion_from_list, node_to_delete))

    return head


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("delete_node_from_list.py",
                                       'delete_node_from_list.tsv',
                                       deletion_from_list_wrapper))
